File: app/src/main/java/com/example/thuctaptotnghiep/data/model/Document.kt
package com.example.thuctaptotnghiep.data.model

import com.google.gson.annotations.SerializedName

data class Document(
    @SerializedName("_id") val id: String,

    // CẬP NHẬT TRỌNG TÂM: Thêm userId để đồng bộ với Backend mới
    @SerializedName("userId") val userId: String? = null,

    @SerializedName("title") val title: String,
    @SerializedName("authorName") val authorName: String,

    @SerializedName("subject") val subject: String? = null,
    @SerializedName("category") val category: String? = null,
    @SerializedName("description") val description: String? = null,
    @SerializedName("tags") val tags: List<String> = emptyList(),

    @SerializedName("fileUrl") val fileUrl: String,
    @SerializedName("size") val size: String? = null,
    @SerializedName("uploadDate") val uploadDate: String,

    @SerializedName("status") val status: String = "pending",
    @SerializedName("downloads") val downloads: Int = 0,
    @SerializedName("views") val views: Int = 0,

    // Các trạng thái tương tác trả về từ Backend
    @SerializedName("isFavorite") val isFavorite: Boolean = false,
    @SerializedName("isWatchLater") val isWatchLater: Boolean = false
)

File: models/Document.js
const mongoose = require('mongoose');

const documentSchema = new mongoose.Schema(
  {
    // Thông tin người đăng
    userId: { type: String, required: true, index: true }, 

    // Thông tin cơ bản của tài liệu
    title: { type: String, required: true, trim: true },
    authorName: { type: String, required: true, trim: true },
    subject: { type: String, required: true }, 
    category: { 
        type: String, 
        enum: ['Slide', 'Đề thi', 'Giáo trình'], 
        required: true 
    }, 
    description: { type: String, trim: true }, 
    tags: { type: [String], default: [] }, 
    
    // Nơi lưu trữ file
    fileUrl: { type: String, required: true },
    thumbnailUrl: { type: String, default: null }, 
    size: { type: String }, 
    
    // Trạng thái (Kiểm duyệt)
    status: { type: String, default: 'pending', enum: ['pending', 'verified', 'failed'] },
    
    // ==========================================
    // [CẬP NHẬT TRỌNG TÂM CHO BẢNG XẾP HẠNG]
    // ==========================================
    
    // Đánh index: true để API lấy bảng xếp hạng chạy siêu tốc
    views: { type: Number, default: 0, index: true },
    downloads: { type: Number, default: 0, index: true },
    
    // [THÊM MỚI] - Chuẩn bị cho chức năng Đánh giá (Rating 1-5 sao)
    averageRating: { type: Number, default: 0 },
    totalRatings: { type: Number, default: 0 },
    
    // ==========================================
    
    // Mảng lưu danh sách UID của những người đã tương tác
    favoritedBy: { type: [String], default: [], index: true },
    watchLaterBy: { type: [String], default: [], index: true },
    
    // (Vẫn giữ uploadDate để tương thích với code cũ của bạn nếu có)
    uploadDate: { type: Date, default: Date.now },
  },
  {
    // Tự động sinh ra createdAt và updatedAt
    timestamps: true,
  }
);

module.exports = mongoose.model('Document', documentSchema);